chalk-solve/src/clauses/clause_visitor.rs

```diff
@@ -0,0 +1,72 @@
+use super::program_clauses::ToProgramClauses;
+use crate::clauses::match_type_kind;
+use crate::DomainGoal;
+use crate::FromEnv;
+use crate::ProgramClause;
+use crate::RustIrDatabase;
+use crate::Ty;
+use chalk_ir::TypeName;
+use rustc_hash::FxHashSet;
+
+pub struct ClauseVisitor<'db, 'set> {
```

Comment:
```suggestion
pub(super) struct ClauseVisitor<'db, 'set> {
```

Reply:
Actually, i'd probably make the type private to this module, and just add a public function as the interface. Something like

```rust
pub(super) fn elaborate_env_clause(
    db: &dyn RustIrDatabase, 
    in_clause: &ProgramClause,
    out: &mut FxHashSet<ProgramClause>,
)
```

It'd be good to add a comment, too. The point of this is -- if we are trying to prove a `FromEnv` goal, then we have to elaborate all the from-env goals in our environment. i.e., if we have `T: Clone` in the environment, that might (indirectly) allow us to prove `T: Copy`. We find that out by adding in the clauses from the `trait Clone`, which include a rule

```rust
FromEnv(T: Copy) :- FromEnv(T: Clone)
```